Plans: auto open for plans descriptions

artpi opened the pull request:
Introduces test mentioned in #6997

I am not sure about this, because there are some issues in the "Business" plan when the tooltip overlays the hover and I am not sure whats the best way to proceed.

## Testing


1. Check out
2. Assign yourself to plansDescriptions: auto test group using env badge
3. Go to /plans. hover on some descriptions.

CC @rralian @lamosty @gwwar @retrofox 

Test live: https://calypso.live/?branch=new/plans-hover-tooltip

lamosty commented:
Could you put some screenshot here? :)

lamosty commented on client/my-sites/plan-features/item.jsx:
```diff
@@ -9,31 +9,48 @@ import classNames from 'classnames';
-		if ( ! description ) {
-			return null;
-		}
+//export default function PlanFeaturesItem( { description, children, className } ) {
```
Why the comment here?

artpi replied:
Whoops, forgot to cleanup :D

artpi replied:
Fixed!

lamosty commented on client/my-sites/plan-features/item.jsx:
```diff
@@ -9,31 +9,48 @@ import classNames from 'classnames';
+		const { description, children, className } = this.props;
+		const classes = classNames( 'plan-features__item', className );
+		let popoverBindings = {};
+		if( abtest( 'plansDescriptions' ) === 'auto' ) {
```
Space after `if` missing. :)

lamosty commented on client/my-sites/plan-features/item.jsx:
```diff
@@ -9,31 +9,48 @@ import classNames from 'classnames';
+			popoverBindings = {
+				onMouseOver: () => this.setState( { showPopover: true } ),
+				onMouseOut: () => this.setState( { showPopover: false } )
+			};
```
JSX props should not use arrow functions because it creates a new function each time you hover the mouse over/out. Better approach is to bind a class method to `this` of the class in `constructor`.

artpi replied:
Fixed, I did as you suggested :)

lamosty commented:
I don't like this. It flickers too much when moving mouse over features and is distracting. [Video](https://drive.google.com/file/d/0B8e-IpRwSdjtVkpiZUZVLTVXUmc/view).

Also, I found a glitch:

![selection_019](https://cloud.githubusercontent.com/assets/4988512/17663941/c01afb16-62f0-11e6-848c-cc42ce303efb.jpg)

Happens sometimes after moving the mouse to one item from the other one.

lamosty commented on client/my-sites/plan-features/item.jsx:
```diff
@@ -9,31 +9,48 @@ import classNames from 'classnames';
+
+	displayName: 'PlanFeaturesItem',
 
+	getInitialState() {
```
We can also write this as

```js

constructor( props ) {
   super( props );
   this.state = { ... };
}
```

gwwar replied:
@artpi did you run into any trouble with binding, that made you use `React.createClass` over an es6 class?

artpi replied:
I again to clean up that piece :/
Sorry, will do ASAP

artpi commented:
> I don't like this. It flickers too much when moving mouse over features and is distracting. Video.

Yes, me neither - I noted  this at the top. I am not sure what really to do with that.

- We can do a bit more code to track state and where the mouse is intering, that would involve more changes to Popover, but @retrofox is refactoring that and I dont know how feasible it is now to change the component
- We can display popover on the right side of the whole business plan column or at the top of current "hivered" field
- We can resign from this whole approach

lamosty commented:
> Yes, me neither - I noted this at the top. 

Yeah, I saw that. :) The issue for me is that it flickers/disturbs too much even on the other, non Business (non Popover hover). Like you move your mouse over features (for better reading, etc) and it shows you the descriptions all the time.

artpi commented:
> The issue for me is that it flickers/disturbs too much even on the other, non Business (non Popover hover).

How would you feel about a small timeout ?

lamosty commented:
> How would you feel about a small timeout ?

We can try that but I'm not sure..it might disconnect action from effect.

gwwar commented:
It may be better to wait for #7217 to land since @retrofox has been doing a lot of work there. Tested this, and the business column seems to be the one glitching the most.

For open on hover we may want behavior similar to http://cherne.net/brian/resources/jquery.hoverIntent.html  which attempts to gauge the user's intent, and doesn't add a number of open/close events if you happen to move your mouse rapidly in and out of the target area.

lamosty commented:
What if we open on hover, but only on hover the (i) icon?

![selection_022](https://cloud.githubusercontent.com/assets/4988512/17673472/8bc3dc58-6321-11e6-80d4-238503d29436.jpg)

That seems more user friendly for me as the user's intent is to see more info.

retrofox commented:
> We can do a bit more code to track state and where the mouse is intering, that would involve more changes to Popover, but @retrofox is refactoring that and I dont know how feasible it is now to change the component

I'd like suggest wait for this PR and then we can improve this weird behavior it it's needed. Let's keep on eye here.

artpi commented:
Thanks @retrofox ! I'll wait :)

retrofox commented:
@artpi Popover refactoring has been merged. Feel free to ping me so I can take a look.

artpi commented:
- I rebased on top of @retrofox work
- I introduced a timout and I think it's pretty smooth now.

The right column when the space is tight is not perfect, but I think its ok.

## [Video](https://cloudup.com/cYo2LPBI4Qe)

lamosty commented:
Yeah, it's a lot better now.

retrofox commented:
@artpi can you try to repro this one?
https://github.com/Automattic/wp-calypso/issues/7518

gwwar commented on client/my-sites/plan-features/item.jsx:
```diff
@@ -9,31 +9,58 @@ import classNames from 'classnames';
+		const classes = classNames( 'plan-features__item', className );
+		let popoverBindings = {};
+		if ( abtest( 'plansDescriptions' ) === 'auto' ) {
+			popoverBindings = {
```
We should be able to avoid creating these bindings on each render.

artpi replied:
I went with @lamosty suggestion to create class methods and bond them in constructor

lamosty replied:
Could we move the abtest check to constructor? If the abtest is set to `auto`, we could assign the `this.mouseOver` and `this.mouseOut` to the particular methods and otherwise set them to noop. I think that could work. :)

gwwar commented:
👍 Looks like its working much better here. I'd prefer if we could fix #7518 before releasing this since the new behavior here makes the issue much more prominent. Also, do you think we could shorten the timeout a bit? It's subjective, but I think we could make it feel a bit more zippy there.

artpi commented:
I shortened the timeout to `100ms` and really think its snappy now :) I'll wait for #7518 and the end of pricing test :)

retrofox commented:
I think with this approach we are forcing to the natural usage of the InfoPopover to another thing. If you have to change the behavior I'd suggest use directly the Popover component instead of this. It has `isVisible` property, `showDelay` property, etc.

lamosty commented on client/my-sites/plan-features/item.jsx:
```diff
@@ -1,39 +1,67 @@
+	}
 
+	renderTipinfo( description ) {
 		return (
```
Could we get the `description` from `this.props`? It's then easier to guess where `description` comes from.

retrofox commented:
In fact it should be a kind of <Tooltip />

retrofox commented:
If we merge #7544 I can rebase this one (if you want).
<img src="http://m.memegen.com/32n62c.jpg" />

artpi commented:
> If we merge #7544 I can rebase this one (if you want).

Yes please! :D I'm onto reviewing #7544 :)

artpi commented:
I vote to close this PR to introduce this test:
https://github.com/Automattic/wp-calypso/pull/7463#issuecomment-242028012
